### PR TITLE
 #160577606 - Authenticate Users

### PR DESCRIPTION
**Pull request by YAHYA-H**
**What does this PR contain?**
This PR returns a JWT to the user upon successfull signup/login.

**Description**
Users once registered via the registration endpoint, should receive a JWT to be required on all subsequent calls to all other endpoints that require authentication.

GIVEN user successfully signs up/registers or logs in
WHEN user passes appropriate JSON object to the register and or login endpoint
THEN user should receive a JWT as the response

**Pivotal Tracker ID**
[#160577606](https://www.pivotaltracker.com/story/show/160577606)

**screenshot**
image

### CODE REVIEW

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
-#
-# from django.conf import settings
-#
-# from rest_framework import authentication, exceptions
-#
-# from .models import User
+import jwt
+import json
+from django.conf import settings
+from rest_framework import authentication, exceptions
```
Comment (esirK) [inline authors/apps/authentication/backends.py:5]: I would suggest you remove unused imports and import them as needed.

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -1,18 +1,20 @@
 import jwt
+import os
```
Comment (esirK) [inline authors/apps/authentication/models.py:5]: The changes in this file do not affect the current task am I right??

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
 from datetime import datetime, timedelta
 
 from django.conf import settings
-from django.contrib.auth.models import (
-    AbstractBaseUser, BaseUserManager, PermissionsMixin
-)
+from django.contrib.auth.models import (AbstractBaseUser, BaseUserManager,
+                                        PermissionsMixin)
 from django.db import models
+from authors.apps.core.token import encode_auth_token
```
Comment (esirK) [inline authors/apps/authentication/models.py]: Is `encode_auth_token` being used in this file??

Comment (YAHYA-H) [reply authors/apps/authentication/models.py]: No, it's not used, I was trying to add token field but I wasn't able

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (YAHYA-H) [reply authors/apps/authentication/models.py:5]: YEAH 😄

Comment (esirK) [reply authors/apps/authentication/models.py:5]: I would suggest you remove it then 😄.

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -11,10 +11,7 @@ class RegistrationSerializer(serializers.ModelSerializer):
     # Ensure passwords are at least 8 characters long, no longer than 128
```
Comment (esirK) [inline authors/apps/authentication/serializers.py]: These changes don't affect this task too.

File: authors/apps/authentication/views.py [hunk 5]
```diff
@@ -1,23 +1,23 @@
...
     # Allow any user (authenticated or not) to hit this endpoint.
-    permission_classes = (AllowAny,)
-    renderer_classes = (UserJSONRenderer,)
+    permission_classes = (AllowAny, )
+    renderer_classes = (UserJSONRenderer, )
     serializer_class = RegistrationSerializer
 
     def post(self, request):
         user = request.data.get('user', {})
+        print("request", request.__dict__)
```
Comment (esirK) [inline authors/apps/authentication/views.py]: What's your view on this statement.

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (YAHYA-H) [reply authors/apps/authentication/serializers.py]: I am not getting @esirK, I have not done any changes to the serializers file. It was yapf package that formatted and brought all these changes. The feature is working, generating token successfully. 🔧 .

File: authors/apps/core/token.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+def encode_auth_token(user_id):
+    """
+    Generates the Auth Token
+    :return: string
+    """
+
+    try:
+        payload = {
+            'exp':
+            datetime.datetime.utcnow() + datetime.timedelta(minutes=60),
```
Comment (gitaumoses4) [inline authors/apps/core/token.py]: I would advice having the time delta as a variable in settings.py.

**Review by gitaumoses4 — COMMENTED** [review]
Kindly ensure you do not autoformat files that are not related to your tasks. This may cause merge conflicts with the rest of the collaborators.

File: authors/apps/core/token.py (hunk 6 above)
Comment (YAHYA-H) [reply authors/apps/core/token.py]: oh yeah thanks for the feedback 👍 😄

Comment (charisschomba) [reply authors/apps/core/token.py]: Nice Progress YAyha......Just make change gitau has recommended.

Comment (SrMoffat) [reply authors/apps/core/token.py]: I have peeped into the master branch (which has the bearbones for the codebase as it was in the very beginning). Based on the current `settings.py` living in `develop`, the specified path for the default authentication class is `authors.apps.authentication.backends.JWTAuthentication` which clearly sets the path to be the `backends.py` file (as indicated in the comments in the snippet). I think either you change the path in the settings.py or the thing will not work. Also, as pointed out, the timedelta could be (os.getenv('TIME_DELTA') then in the .env have something like TIME_DELTA=60

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (esirK) [reply authors/apps/authentication/serializers.py]: I get what happened. I would still ask you to review the files you `git add`. Only add files related to the task at hand.

Comment (esirK) [reply authors/apps/authentication/serializers.py]: Kindly remove it since it doesn't correspond to this task.

File: authors/apps/authentication/models.py [hunk 7]
```diff
@@ -1,18 +1,16 @@
 import jwt
```
Comment (esirK) [inline authors/apps/authentication/models.py]: Also, remove this file as @gitaumoses4 has pointed out.

File: authors/apps/authentication/views.py [hunk 8]
```diff
@@ -1,24 +1,24 @@
 from rest_framework import status
-from rest_framework.generics import RetrieveUpdateAPIView,CreateAPIView
+from rest_framework.generics import RetrieveUpdateAPIView, CreateAPIView
```
Comment (esirK) [inline authors/apps/authentication/views.py]: same applies for this file too.

File: authors/urls.py [hunk 9]
```diff
@@ -18,12 +18,10 @@
...
 from rest_framework_swagger.views import get_swagger_view
 
-schema_view = get_swagger_view(title='Authors Haven API')
-
-
 urlpatterns = [
     path('admin/', admin.site.urls),
-    path('api/', include('authors.apps.authentication.urls', namespace='authentication')),
-    path('', schema_view),
+    path(
```
Comment (esirK) [inline authors/urls.py]: also for this one.

Comment (YAHYA-H) [reply authors/urls.py]: I'll revert back to the previous commit and redo the task because there is a lot of files that was edited

File: authors/apps/authentication/backends.py [hunk 10]
```diff
@@ -1,13 +1,49 @@
...
+
+    def authenticate(self, request):
+        request.user = None
+
+        # returns Authorization header
+        # as a bytestring
+        auth_header = authentication.get_authorization_header(request).split()
+
+        if not auth_header or auth_header[0].lower() != self.keyword.lower():
+            return None
```
Comment (charisschomba) [inline authors/apps/authentication/backends.py:27]: Return message is no descriptive.

File: authors/apps/authentication/backends.py [hunk 11]
```diff
@@ -1,13 +1,49 @@
...
+        elif len(auth_header) > 2:
+            raise exceptions.AuthenticationFailed('Invalid token header. Token string should not contain spaces.')
+
+        return self.authenticate_credentials(request, token)
+
+    def authenticate_credentials(self, request, token):
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except:
+            raise exceptions.AuthenticationFailed('Cannot decode token.')
```
Comment (charisschomba) [inline authors/apps/authentication/backends.py]: This message could be improved

File: authors/apps/authentication/backends.py [hunk 12]
```diff
@@ -1,13 +1,49 @@
...
+
+from django.conf import settings
+
+from rest_framework import authentication, exceptions
+
+from .models import User
+
+
+class JWTAuthentication(authentication.BaseAuthentication):
+    """ This validate the token."""
```
Comment (charisschomba) [inline authors/apps/authentication/backends.py]: Doc string could be more descriptive.

**Comment by YAHYA-H** [comment]
why is there 7 files changed? I think you haven't pull the changes I did @imireallan

**Comment by esirK** [comment]
The *Pivotal Tracker ID* doesn't reflect the id of this story.

File: authors/apps/authentication/renderers.py [hunk 13]
```diff
@@ -18,6 +19,11 @@ def render(self, data, media_type=None, renderer_context=None):
             # rendering errors.
             return super(UserJSONRenderer, self).render(data)
```
Comment (esirK) [inline authors/apps/authentication/renderers.py]: in python 3, there is a better way to call super. Check it out and update this part of the code.

Comment (YAHYA-H) [reply authors/apps/authentication/renderers.py]: 😄  🆗

File: authors/apps/core/token.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import os
+import datetime
+from datetime import timedelta
+import jwt
+from authors.settings import SECRET_KEY
+
+def encode_auth_token(user_id):
```
Comment (esirK) [inline authors/apps/core/token.py]: This could be placed in the `backends` file.

**Comment by esirK** [comment]
~~As stated on `PT dev notes`, **The app should no longer use sessions, but should use the newly implemented JWT strategy for authentication.** You are required to do the actual implementation of the `authentication` method of the `JWTAuthentication` class. @YAHYA-H~~

File: authors/apps/authentication/backends.py [hunk 15]
```diff
@@ -1,13 +1,50 @@
...
+        try:
+            user = User.objects.get(pk=payload['id'])
+        except User.DoesNotExist:
+            exceptions.AuthenticationFailed('No user Found')
+        if not user.is_active:
+            raise exceptions.AuthenticationFailed('User has been deactivated')
+        if not user.is_verified:
+            raise exceptions.AuthenticationFailed('User has not been verified')
+
+        return(user, token)
```
Comment (esirK) [inline authors/apps/authentication/backends.py]: the brackets aren't necessary.

File: authors/apps/authentication/models.py (hunk 7 above)
Comment (esirK) [reply authors/apps/authentication/models.py]: This wasn't done.

File: authors/apps/authentication/backends.py (hunk 12 above)
Comment (esirK) [reply authors/apps/authentication/backends.py]: @YAHYA-H Did you ignore the above comment??

File: authors/apps/authentication/backends.py [hunk 16]
```diff
@@ -1,13 +1,50 @@
...
+        auth_header = authentication.get_authorization_header(request).split()
+
+        if not auth_header or auth_header[0].lower() != self.keyword.lower():
+            return None
+        if len(auth_header) == 1:
+                raise exceptions.AuthenticationFailed('Invalid token header. No credentials provided.')
+        elif len(auth_header) > 2:
+            raise exceptions.AuthenticationFailed('Invalid token header. Token string should not contain spaces.')
+
+        return self.authenticate_credentials(request, token)
```
Comment (esirK) [inline authors/apps/authentication/backends.py]: Where is this `token` variable declared??

File: authors/apps/authentication/backends.py [hunk 17]
```diff
@@ -1,13 +1,50 @@
...
+        successful, return the user and token. If not, throw an error.
+        """
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except:
+            raise exceptions.AuthenticationFailed('Cannot decode token.')
+        try:
+            user = User.objects.get(pk=payload['id'])
+        except User.DoesNotExist:
+            exceptions.AuthenticationFailed('No user Found')
```
Comment (esirK) [inline authors/apps/authentication/backends.py]: Aren't you missing a `raise` statement here??

File: authors/apps/authentication/backends.py [hunk 18]
```diff
@@ -1,13 +1,50 @@
...
+        return self.authenticate_credentials(request, token)
+
+    def authenticate_credentials(self, request, token):
+        """
+        authenticate the given credentials. If authentication is
+        successful, return the user and token. If not, throw an error.
+        """
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except:
```
Comment (esirK) [inline authors/apps/authentication/backends.py]: This should be more explicit. What exception are you expecting? This should raise a warning on your `IDE`

**Comment by esirK** [comment]
On testing this out, seems like the `authenticate` method of the `JWTAuthentication` class is not working. Kindly test out the `/user/` endpoint with a valid token to see if this method works.

File: authors/apps/authentication/backends.py (hunk 10 above)
Comment (imireallan) [reply authors/apps/authentication/backends.py:27]: please give me suggestions on this

File: authors/apps/authentication/backends.py (hunk 16 above)
Comment (imireallan) [reply authors/apps/authentication/backends.py]: the token variable is declared in the `authenticate_credentials method`

File: authors/apps/authentication/backends.py (hunk 18 above)
Comment (imireallan) [reply authors/apps/authentication/backends.py]: let me look at it now..

File: authors/apps/authentication/backends.py (hunk 10 above)
Comment (esirK) [reply authors/apps/authentication/backends.py:27]: This is alright. The return statement should be None

File: authors/apps/authentication/backends.py (hunk 16 above)
Comment (esirK) [reply authors/apps/authentication/backends.py]: Kindly re-check this. It is being passed as a parameter to `authenticate_credentials` method

File: authors/apps/authentication/backends.py (hunk 12 above)
Comment (YAHYA-H) [reply authors/apps/authentication/backends.py]: thanks, I'll implement it, Allan was working on this class 👍 😄

File: authors/settings.py [hunk 19]
```diff
@@ -158,7 +158,10 @@
 REST_FRAMEWORK = {
     'EXCEPTION_HANDLER': 'authors.apps.core.exceptions.core_exception_handler',
     'NON_FIELD_ERRORS_KEY': 'error',
-
+    'DEFAULT_PERMISSION_CLASSES': (
```
Comment (esirK) [inline authors/settings.py]: I don't think we need this.

File: requirements.txt [hunk 20]
```diff
@@ -1,15 +1,24 @@
+certifi==2018.8.24
```
Comment (esirK) [inline requirements.txt]: Kindly remove all packages that are not related to this task.

**Comment by esirK** [comment]
Looks good. Work on the comments and it will be good for merging.

File: authors/settings.py (hunk 19 above)
Comment (imireallan) [reply authors/settings.py]: the default classes i don't think we need them too..

File: authors/apps/authentication/backends.py [hunk 21]
```diff
@@ -1,14 +1,53 @@
...
-    def authenticate(self, another):
-        pass
+from .models import User
+
+
+class JWTAuthentication(authentication.BaseAuthentication):
+    """ Authenticate a received token from the 'Authorization' Headers by
+    the keyword 'Token' """
+
+    # initializing the token as an empty string
```
Comment (imireallan) [inline authors/apps/authentication/backends.py:15]: try and refactor this comment

File: authors/apps/authentication/backends.py [hunk 22]
```diff
@@ -1,14 +1,53 @@
...
+    keyword = 'Token'
+
+    def authenticate(self, request):
+        request.user = None
+
+        # returns Authorization header
+        # as a bytestring
+
+        auth_header = authentication.get_authorization_header(request).split()
+        if not auth_header:
```
Comment (imireallan) [inline authors/apps/authentication/backends.py]: line 24 and 27 doing the same thing - they both check for the absence of auth_header

File: authors/apps/authentication/backends.py [hunk 23]
```diff
@@ -1,14 +1,53 @@
...
+        return self.authenticate_credentials(request, auth_header[1].decode())
+
+    def authenticate_credentials(self, request, token):
+        """
+        authenticate the given credentials. If authentication is
+        successful, return the user and token. If not, throw an error.
+        """
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except jwt.DecodeError:
```
Comment (imireallan) [inline authors/apps/authentication/backends.py]: you can also add an exception that handles the expiry of the token-
`except jwt.ExpiredTokenSignature
       raise exceptions.AuthenticationFailed`

**Review by YAHYA-H — COMMENTED** [review]
@esirK and @imireallan , for default permission class, this is important for articles, to ensure that authorized users only have visibility to our feature. We control what user can view,...http://www.django-rest-framework.org/api-guide/permissions/#object-level-permissions

**Comment by imireallan** [comment]
@YAHYA-H the permissions can be put on the article routes as it is currently done in the user routes
like you can import the permissions on `from rest_framework.permissions import AllowAny, IsAuthenticated` and then you instantiate the permissions on the articles views

**Comment by esirK** [comment]
@YAHYA-H  you don't need to include this. Once you add it to the permissions field that's done. Remember, Explicit is better than implicit.

**Comment by esirK** [comment]
kindly rebase to develop and run flake8 before pushing.

File: authors/settings.py [hunk 24]
```diff
@@ -158,7 +158,10 @@
 REST_FRAMEWORK = {
     'EXCEPTION_HANDLER': 'authors.apps.core.exceptions.core_exception_handler',
     'NON_FIELD_ERRORS_KEY': 'error',
-
+    # 'DEFAULT_PERMISSION_CLASSES': (
```
Comment (esirK) [inline authors/settings.py]: what does this comment tell us about the code??

Comment (YAHYA-H) [reply authors/settings.py]: i've removed that last night

File: authors/apps/authentication/backends.py (hunk 22 above)
Comment (esirK) [reply authors/apps/authentication/backends.py]: did you resolve this??

File: authors/apps/authentication/backends.py [hunk 25]
```diff
@@ -1,14 +1,55 @@
...
+
+    def authenticate_credentials(self, request, token): # NOQA
+        """
+        authenticate the given credentials. If authentication is
+        successful, return the user and token. If not, throw an error.
+        """
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except (jwt.DecodeError, jwt.ExpiredSignatureError) as e:
+            raise exceptions.AuthenticationFailed(str(e))
```
Comment (esirK) [inline authors/apps/authentication/backends.py]: This isn't what I advised you to do @YAHYA-H . The raised exception message won't be descriptive.

**Comment by esirK** [comment]
Looks good now. Good Job everyone.
